bytebuffer - Add proper overloads for functions to reflect returned types

murbanowicz opened the pull request:
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [ ] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/protobufjs/bytebuffer.js/wiki/API#bytebufferreadistringoffset

Returned types for changed functions are strictly related to the functions signatures, so I did add them, so there is no need to cast the values anymore as the returned types with omitted or not omited offset are predictable.

murbanowicz commented:
@dcappellin

dcappellin commented:
OK

murbanowicz commented:
@dcappellin you have to approve changes in review part to make it pass automatically. Could you?

dcappellin commented on types/bytebuffer/index.d.ts:
```diff
@@ -3,6 +3,7 @@
 // Definitions by: Denis Cappellin <https://github.com/cappellin>
```
My GitHub address is bad, cause it misses a "d". Can you fix it, please? 
    
    // Definitions by: Denis Cappellin <https://github.com/dcappellin>

murbanowicz replied:
I would say to make it in separate PR. I have spotted it as well.

murbanowicz commented:
@andrewbranch - Sorry for pinging you but you're the first on the list.
We have a strange situation here. Maintainer has changed the github address/nickname and it makes it impossible to automerge it.
Could you please help with that?

murbanowicz commented on types/bytebuffer/index.d.ts:
```diff
@@ -316,7 +317,8 @@ declare class ByteBuffer
      */
-    readCString( offset?: number ): string;
+    readCString(): string;
+    readCString( offset: number ): { string: string, length: number};
```
I cannot agree.
Please take a look into docs.
` { string: string, length: number}` is returned when and only when offset is provided.
When there is no offset provided, then just `string` is returned.

My change make typing exactly as they are in the runtime.
If it brakes something to someone, it means it was broken before.
Having `string | { string: string, length: number}` is bad practice, because you then have to cast later on to expected result and that is exactly why I came here to fix it because we have plenty of castings in our codebase around this functions.

dcappellin replied:
OK, clear. I wrote the definitions years ago and I need some info to understand. What about these methods:

 writeVString( str: string, offset?: number ): ByteBuffer | number;
    /**
     * Writes a 32bit base 128 variable-length integer.
     */
    writeVarint32( value: number, offset?: number ): ByteBuffer | number;
    /**
     * Writes a zig-zag encoded 32bit base 128 variable-length integer.
     */
    writeVarint32ZigZag( value: number, offset?: number ): ByteBuffer | number;
    /**
     * Writes a 64bit base 128 variable-length integer.
     */
    writeVarint64( value: number | Long, offset?: number ): ByteBuffer;
    /**
     * Writes a zig-zag encoded 64bit base 128 variable-length integer.
     */
    writeVarint64ZigZag( value: number | Long, offset?: number ): ByteBuffer | number;

Do we need to split the returning value based on signature or are they just fine as they are?

murbanowicz replied:
Cool.
I thought I've found all of these issues but looks like there is more.
I will do that either in this PR or next one.
It's a shame there is this stupid issue with your username and your approval is not treated as maintainer's :/

dcappellin replied:
Please fix those methods in this PR and then I’ll approve.

murbanowicz replied:
Done! Please take a look and approve.

murbanowicz commented:
@dcappellin 
1. I have fixed overloads for functions which you've found not updated.
2. I have fixed your Github name in headers.
3. I did add simple test to cover it.

murbanowicz commented:
@RyanCavanaugh @RyanCavanaugh @kieferrm @blakeembrey @Steve-Fenton @johnnyreilly @basarat
Can any of you help with the issue which we have here - maintainer approved, but there is wrong GitHub name in the file, so he is not recognized as maintainer?
This PR fixes this issue as well, so we should not need your help in the future, but we need it now...

murbanowicz commented:
@dcappelin: I had to push again after prettier as VS Code ruined formatting, so it couldn't be reviewed easily anymore as commit was changing whole file. Please approve again.

murbanowicz commented:
Ready to merge
